### PR TITLE
KFSPTS-26091 add locking document ID to cache

### PR DESCRIPTION
**Pull request by jhulslander**

### CODE REVIEW

**Comment by jhulslander** [comment]
This is ready for review again

File: src/main/java/edu/cornell/kfs/coa/document/CuAccountGlobalMaintainableImpl.java
Comment (cah292) [inline src/main/java/edu/cornell/kfs/coa/document/CuAccountGlobalMaintainableImpl.java:55]: I'm wondering if the cache entry also needs to be cleared when the Account Global adds or removes collection lines, if this current mechanism is not enough to perform such clearing. If such an addition is needed, perhaps the `processAfterPost()` method could potentially be used for such entry clearing when it detects an add-line or remove-line operation, both single-valued and/or multi-valued.

File: src/main/java/edu/cornell/kfs/coa/document/CuSubObjectCodeMaintainableImpl.java
Comment (cah292) [inline src/main/java/edu/cornell/kfs/coa/document/CuSubObjectCodeMaintainableImpl.java:24]: You should also add this clear-all-entries logic to the Account Reversion maintainable.

File: src/main/resources/edu/cornell/kfs/coa/document/datadictionary/SubObjectCodeMaintenanceDocument.xml
Comment (cah292) [inline src/main/resources/edu/cornell/kfs/coa/document/datadictionary/SubObjectCodeMaintenanceDocument.xml:6]: Does this file need the KualiCo license header? It's not overlaying the related file from base code.

File: src/main/resources/kfs.ehcache.xml
Comment (cah292) [inline src/main/resources/kfs.ehcache.xml:76]: You may want to specify an explicit time-to-live value for this cache.

**Review by cah292 — CHANGES_REQUESTED** [review]
I found a few areas that might need more revisions, but overall the changes look good.

File: src/main/java/edu/cornell/kfs/coa/document/CuAccountGlobalMaintainableImpl.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/coa/document/CuAccountGlobalMaintainableImpl.java:55]: I am confused at what you're saying here.  Should this clear cache logic be moved to processAfterPost()?  If so, should this logic be moved to processAfterPost() in the other maintainable classes?

File: src/main/resources/kfs.ehcache.xml
Comment (jhulslander) [reply src/main/resources/kfs.ehcache.xml:76]: OK, I will set timeToLiveSeconds="1800", 30 minutes seems reasonable for this particular cache

File: src/main/java/edu/cornell/kfs/coa/document/CuSubObjectCodeMaintainableImpl.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/coa/document/CuSubObjectCodeMaintainableImpl.java:24]: OK, added code to AccountReversionMaintainableImpl

File: src/main/java/edu/cornell/kfs/coa/document/CuAccountGlobalMaintainableImpl.java
Comment (cah292) [reply src/main/java/edu/cornell/kfs/coa/document/CuAccountGlobalMaintainableImpl.java:55]: If the user edits the Account Global doc between submission attempts, that may affect what downstream objects need maintenance locks during routing. This spot will at least clear the cache when the locks are being persisted to the database, but if this spot doesn't affect adding/removing accounts from the global doc, then appropriate cache clearing might need to be added to `processAfterPost()` or a similar hook.

The other maintainables don't need to override the `processAfterPost()` logic, since they only need to clear the cache when they store or delete locks that may affect Account Global. However, after considering this a little further, the other maintainables may want to move the cache clearing from their `generateMaintenanceLocks()` methods to their `doRouteStatusChange()` methods, so that the cache gets cleared when the objects start or finish their routing.

Similarly, it might make sense for the Account Global maintainable to move its single-entry clearing from `generateMaintenanceLocks()` to `doRouteStatusChange()` for clarity, unless that messes up the new caching feature.

**Comment by jhulslander** [comment]
I've mad some additional changes based on the last comments.  I want to test these changes in prekew before they merged into develop.  If you want to review again, please do so, but hold off on merging till I validate  some more.

File: src/main/java/edu/cornell/kfs/coa/document/AccountReversionMaintainableImpl.java
Comment (cah292) [inline src/main/java/edu/cornell/kfs/coa/document/AccountReversionMaintainableImpl.java:267]: I noticed that the various `doRouteStatusChange()` overrides are set up to clear the cache on every status change. During your Pre-KEW testing, you may want to verify whether the cache clearing should be limited to certain specific status changes (ENROUTE, PROCESSED, DISAPPROVED, etc.).

**Review by cah292 — COMMENTED** [review]
I did a quick scan of the revised code, and noticed something that may or may not need further attention. I will perform a more thorough review tomorrow.

**Review by cah292 — COMMENTED** [review]
I noticed one spot that may or may not need revising, but otherwise the latest revisions look good. Also, the upcoming Pre-KEW testing should confirm whether additional cache clearing is needed when the Account Global doc adds or removes collection lines.

File: src/main/java/org/kuali/kfs/krad/maintenance/MaintenanceUtils.java
Comment (jhulslander) [reply src/main/java/org/kuali/kfs/krad/maintenance/MaintenanceUtils.java:130]: I can change it

**Comment by jhulslander** [comment]
I did run into a problem with adding detail lines.  I've added "clear cache" logic to addMultipleValueLookupResults() and addNewLineToCollection().  I need to add "clear cache" to removing a detail line.  I haven't found a.good approach to doing that yet

**Comment by jhulslander** [comment]
I believe the bug with the approver adding removing details was fixed with clearing cache on prepareForSave().  This is ready for review again

File: src/main/java/edu/cornell/kfs/coa/document/CuAccountGlobalMaintainableImpl.java
Comment (cah292) [inline src/main/java/edu/cornell/kfs/coa/document/CuAccountGlobalMaintainableImpl.java:75]: Can you please confirm if you're still seeing the expected performance gains locally or in Pre-KEW? If you're putting a cache-clearing operation in the `prepareForSave()` method, I'm wondering if that would cause the cache to be cleared too frequently and potentially nullify the intended performance improvements.

**Review by cah292 — COMMENTED** [review]
I had a question about the placement of one of the cache-clearing calls, but otherwise the revised changes look good.

File: src/main/java/edu/cornell/kfs/coa/document/CuAccountGlobalMaintainableImpl.java
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/coa/document/CuAccountGlobalMaintainableImpl.java:75]: my pre-kew testing has good performance results

Comment (cah292) [reply src/main/java/edu/cornell/kfs/coa/document/CuAccountGlobalMaintainableImpl.java:75]: Thanks for confirming. I'm okay with moving the changes forward to DEV for more in-depth validation.

**Review by cah292 — APPROVED** [review]
Changes look good, unit tests pass and my concerns have been addressed. Merging.

**Pull request merged**
